### PR TITLE
feat: restrict effectors

### PR DESCRIPTION
**Pull request by kmd-fl**
## Description
[Provide a brief description of the changes introduced by this pull request and the motivation behind them.]

## Motivation
safe nox

## Proposed Changes
- generate configs inside nox
- add new method to add module that doesn't accept config
- other methods still accept it, but use only the name 
- make curl (from the config example) a default effector

## Additional Notes
- System services still use the config from distro
- Marine configs are still stored when we add module distro (TODO: fix it)
- `allowd_binaries` in the peer info API now prints the list of paths used in effectors

Config example:
```toml
[effectors.curl]
wasm_cid = "bafkreids22lgia5bqs63uigw4mqwhsoxvtnkpfqxqy5uwyyerrldsr32ce"
  [effectors.curl.allowed_binaries]
  curl = "/usr/bin/curl"
```
or
```toml
[effectors.curl]
wasm_cid = "bafkreids22lgia5bqs63uigw4mqwhsoxvtnkpfqxqy5uwyyerrldsr32ce"
allowed_binaries = { curl = "/usr/bin/curl" }
```

New api
```aqua
service Dist("dist"):
   add_module_bytes_from_vault(name: string, module_cid: string) -> string
```

### CODE REVIEW

File: nox/src/node.rs [hunk 1]
```diff
@@ -310,7 +304,8 @@ impl<RT: AquaRuntime> Node<RT> {
             node_version: env!("CARGO_PKG_VERSION"),
             air_version: air_interpreter_wasm::VERSION,
             spell_version: spell_version.clone(),
-            allowed_binaries,
+            // TODO: remove
+            allowed_binaries: vec![],
```
Comment (justprosh) [inline nox/src/node.rs:308]: Maybe put cids here for now

File: particle-modules/src/modules.rs [hunk 2]
```diff
@@ -67,33 +68,53 @@ impl ModuleRepository {
...
-        if let Some(binaries) = &config.config.mounted_binaries {
-            for requested_binary in binaries.values() {
-                if let Some(requested_binary) = requested_binary.as_str() {
-                    let requested_binary_path = Path::new(requested_binary);
-                    if !self.allowed_binaries.contains(requested_binary_path) {
-                        return Err(ForbiddenMountedBinary {
-                            forbidden_path: requested_binary.to_string(),
+    pub fn add_module(&self, name: String, module: Vec<u8>) -> Result<Hash> {
+        // TODO: remove unwrap
+        let hash = Hash::new(&module).unwrap();
```
Comment (justprosh) [inline particle-modules/src/modules.rs:77]: It's time to remove it

File: particle-modules/src/modules.rs [hunk 3]
```diff
@@ -67,33 +68,53 @@ impl ModuleRepository {
...
 
-    pub fn add_module(&self, module: Vec<u8>, config: TomlMarineNamedModuleConfig) -> Result<Hash> {
-        // TODO: remove unwrap
+    // TODO: generate config for modules also
+    pub fn add_system_module(
+        &self,
+        module: Vec<u8>,
+        config: TomlMarineNamedModuleConfig,
+    ) -> Result<Hash> {
         let hash = Hash::new(&module).unwrap();
```
Comment (justprosh) [inline particle-modules/src/modules.rs:116]: And here

**Review by justprosh — APPROVED** [review]

File: nox/src/node.rs (hunk 1 above)
Comment (kmd-fl) [reply nox/src/node.rs:310]: yeah, but I'm not sure if it will break some Tolya's scripts rn

Comment (kmd-fl) [reply nox/src/node.rs]: fixed

File: crates/server-config/src/services_config.rs [hunk 4]
```diff
@@ -62,21 +63,27 @@ impl ServicesConfig {
...
-                };
-                path.to_path_buf()
+            .map(|(cid, effector)| {
+                let effector = effector
+                    .into_iter()
+                    .map(|(name, path_str)| {
+                        let path = Path::new(&path_str);
+                        match path.try_exists() {
+                            Err(err) => log::warn!("cannot check effector `{path_str}`: {err}"),
+                            Ok(false) => log::warn!("effector `{path_str}` does not exist"),
```
Comment (folex) [inline crates/server-config/src/services_config.rs:80]: Do you mean "binary {path_str} does not exist"? Or "executable".

cuz 1 effector can have any number of binaries, and effector is referenced by CID, not by binary's name

File: particle-modules/src/modules.rs [hunk 5]
```diff
@@ -67,33 +68,53 @@ impl ModuleRepository {
...
-    fn check_module_mounted_binaries(&self, config: &TomlMarineNamedModuleConfig) -> Result<()> {
-        if let Some(binaries) = &config.config.mounted_binaries {
-            for requested_binary in binaries.values() {
-                if let Some(requested_binary) = requested_binary.as_str() {
-                    let requested_binary_path = Path::new(requested_binary);
-                    if !self.allowed_binaries.contains(requested_binary_path) {
-                        return Err(ForbiddenMountedBinary {
-                            forbidden_path: requested_binary.to_string(),
+    pub fn add_module(&self, name: String, module: Vec<u8>) -> Result<Hash> {
+        // TODO: remove unwrap
```
Comment (folex) [inline particle-modules/src/modules.rs:76]: ?

File: particle-modules/src/modules.rs [hunk 6]
```diff
@@ -67,33 +68,53 @@ impl ModuleRepository {
...
+        let hash = Hash::new(&module)?;
+        let (logger_enabled, mounted) = Self::get_module_effects(&module)?;
+        let mut effector_settings = None;
+        if !mounted.is_empty() {
+            if let Some(binaries) = self.allowed_effectors.get(&hash) {
+                // check that all configured binaries are the same as in the module
```
Comment (folex) [inline particle-modules/src/modules.rs:82]: why?

File: crates/server-config/src/services_config.rs (hunk 4 above)
Comment (kmd-fl) [reply crates/server-config/src/services_config.rs:80]: Yes, I meant binary. 
Need to fix to smth like `binary {path_str} of effector {CID} does not exist`

File: crates/server-config/src/node_config.rs [hunk 7]
```diff
@@ -365,6 +382,8 @@ pub struct NodeConfig {
 
     pub allowed_binaries: Vec<String>,
 
+    pub allowed_effectors: HashMap<Hash, HashMap<String, String>>,
```
[the remark this answers is not in the document]
Comment (kmd-fl) [reply crates/server-config/src/node_config.rs:385]: Not yet. Will add a flag fir local development in the next PR

File: particle-modules/src/modules.rs (hunk 5 above)
Comment (kmd-fl) [reply particle-modules/src/modules.rs:76]: Forgot to remove. I 'll fix it.

**Pull request merged**